File: frontend/lib/services/autonomous-agents.ts
// Partie 23 -- autonomous agents API client, api/routers/autonomous_agents.py's
// real surface.

import { api } from "@/lib/api";

export type AutonomousAgentStatus = "idle" | "planning" | "executing" | "paused" | "completed" | "error";

export interface AutonomousAgent {
  id: string;
  organization_id: string;
  created_by: string | null;
  name: string;
  description: string | null;
  goal: string;
  status: AutonomousAgentStatus;
  max_steps: number;
  current_step: number;
  tools_enabled: { name: string; enabled: boolean; config?: Record<string, unknown> }[];
  guardrails: Record<string, unknown>;
  memory_config: Record<string, unknown>;
  error: string | null;
  created_at: string;
  updated_at: string;
}

export interface AutonomousAgentListResponse {
  items: AutonomousAgent[];
  total: number;
  limit: number;
  offset: number;
}

export interface AgentStatus {
  status: AutonomousAgentStatus;
  current_step: number;
  max_steps: number;
  error: string | null;
}

export interface AgentPlan {
  id: string;
  agent_id: string;
  goal: string;
  steps: { description: string; depends_on: number[] }[];
  status: "pending" | "running" | "completed" | "failed";
  created_at: string;
}

export interface AgentStep {
  id: string;
  plan_id: string;
  step_number: number;
  action: string;
  parameters: Record<string, unknown> | null;
  result: Record<string, unknown> | null;
  status: "pending" | "running" | "completed" | "failed" | "skipped";
  error: string | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface AgentMemory {
  id: string;
  agent_id: string;
  memory_type: "short_term" | "long_term" | "episodic";
  content: string;
  importance: number;
  created_at: string;
}

export interface AgentCollaboration {
  id: string;
  initiator_agent_id: string;
  collaborator_agent_id: string;
  task: string;
  status: "pending" | "accepted" | "running" | "completed" | "failed";
  result: Record<string, unknown> | null;
  created_at: string;
}

export const listAutonomousAgents = (orgId: string, limit = 50, offset = 0) =>
  api.get<AutonomousAgentListResponse>(`/organizations/${orgId}/autonomous-agents?limit=${limit}&offset=${offset}`);

export const createAutonomousAgent = (orgId: string, data: {
  name: string; description?: string; goal: string; max_steps?: number;
  tools_enabled?: Record<string, unknown>[]; guardrails?: Record<string, unknown>; memory_config?: Record<string, unknown>;
}) => api.post<AutonomousAgent>(`/organizations/${orgId}/autonomous-agents`, data);

export const getAutonomousAgent = (id: string) => api.get<AutonomousAgent>(`/autonomous-agents/${id}`);

export const updateAutonomousAgent = (id: string, data: Partial<Pick<AutonomousAgent, "name" | "description" | "goal" | "max_steps" | "tools_enabled" | "guardrails" | "memory_config">>) =>
  api.patch<AutonomousAgent>(`/autonomous-agents/${id}`, data);

export const deleteAutonomousAgent = (id: string) => api.delete(`/autonomous-agents/${id}`);

export const runAutonomousAgent = (id: string) => api.post<AutonomousAgent>(`/autonomous-agents/${id}/run`);
export const pauseAutonomousAgent = (id: string) => api.post<AutonomousAgent>(`/autonomous-agents/${id}/pause`);
export const resumeAutonomousAgent = (id: string) => api.post<AutonomousAgent>(`/autonomous-agents/${id}/resume`);
export const stopAutonomousAgent = (id: string) => api.post<AutonomousAgent>(`/autonomous-agents/${id}/stop`);
export const getAgentStatus = (id: string) => api.get<AgentStatus>(`/autonomous-agents/${id}/status`);

export const listAgentPlans = (id: string) => api.get<AgentPlan[]>(`/autonomous-agents/${id}/plans`);
export const getAgentPlan = (id: string, planId: string) => api.get<AgentPlan>(`/autonomous-agents/${id}/plans/${planId}`);
export const listAgentSteps = (id: string, planId: string) => api.get<AgentStep[]>(`/autonomous-agents/${id}/plans/${planId}/steps`);

export const getAgentMemory = (id: string, memoryType?: string) =>
  api.get<AgentMemory[]>(`/autonomous-agents/${id}/memory${memoryType ? `?memory_type=${memoryType}` : ""}`);
export const addAgentMemory = (id: string, data: { content: string; memory_type?: string; importance?: number }) =>
  api.post<AgentMemory>(`/autonomous-agents/${id}/memory`, data);
export const deleteAgentMemory = (id: string, memoryId: string) => api.delete(`/autonomous-agents/${id}/memory/${memoryId}`);

export const collaborateAgents = (id: string, data: { collaborator_agent_id: string; task: string }) =>
  api.post<AgentCollaboration>(`/autonomous-agents/${id}/collaborate`, data);
export const listCollaborations = (id: string) => api.get<AgentCollaboration[]>(`/autonomous-agents/${id}/collaborations`);

File: api/routers/autonomous_agents.py
"""Partie 23 -- autonomous agent endpoints. Org-scoped list/create
routes reuse `require_org_member`/`require_org_admin` directly;
single-resource routes use `require_autonomous_agent_member`/
`require_autonomous_agent_admin` (api/security/autonomous_agents.py),
a flat `/autonomous-agents/{id}` path -- same access-level adaptation
already applied for Parties 19-22."""

import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import get_current_user, get_db
from api.models.autonomous_agent import AutonomousAgent
from api.models.organization import OrganizationMember
from api.models.user import User
from api.schemas.autonomous_agent import (
    AgentCollaborationRequest, AgentCollaborationResponse, AgentCostResponse, AgentMemoryCreateRequest, AgentMemoryResponse,
    AgentPlanResponse, AgentStatusResponse, AgentStepResponse, AutonomousAgentCreateRequest, AutonomousAgentListResponse,
    AutonomousAgentResponse, AutonomousAgentUpdateRequest,
)
from api.security.autonomous_agents import require_autonomous_agent_admin, require_autonomous_agent_member
from api.security.organizations import require_org_admin, require_org_member
from api.services import autonomous_agents as autonomous_agents_service
from api.tasks.autonomous_agents import schedule_autonomous_agent_run

router = APIRouter(tags=["autonomous-agents"])


@router.get("/organizations/{org_id}/autonomous-agents", response_model=AutonomousAgentListResponse)
async def list_autonomous_agents_endpoint(
    org_id: uuid.UUID, limit: int = Query(default=50, ge=1, le=200), offset: int = Query(default=0, ge=0),
    _caller: OrganizationMember = Depends(require_org_member), db: AsyncSession = Depends(get_db),
):
    return await autonomous_agents_service.list_autonomous_agents(db, org_id, limit, offset)


@router.post("/organizations/{org_id}/autonomous-agents", response_model=AutonomousAgentResponse, status_code=status.HTTP_201_CREATED)
async def create_autonomous_agent_endpoint(
    org_id: uuid.UUID, payload: AutonomousAgentCreateRequest, _caller: OrganizationMember = Depends(require_org_admin),
    current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db),
):
    agent = await autonomous_agents_service.create_autonomous_agent(db, org_id, payload.model_dump(), current_user.id)
    await db.commit()
    await db.refresh(agent)
    return agent


@router.get("/autonomous-agents/{agent_id}", response_model=AutonomousAgentResponse)
async def get_autonomous_agent_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member)):
    agent, _caller = agent_ctx
    return agent


@router.patch("/autonomous-agents/{agent_id}", response_model=AutonomousAgentResponse)
async def update_autonomous_agent_endpoint(
    payload: AutonomousAgentUpdateRequest, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_admin),
    db: AsyncSession = Depends(get_db),
):
    agent, _caller = agent_ctx
    updated = await autonomous_agents_service.update_autonomous_agent(db, agent.id, payload.model_dump(exclude_unset=True))
    await db.commit()
    await db.refresh(updated)
    return updated


@router.delete("/autonomous-agents/{agent_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_autonomous_agent_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_admin), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    await autonomous_agents_service.delete_autonomous_agent(db, agent.id)
    await db.commit()


# --------------------------------------------------------------- execution

@router.post("/autonomous-agents/{agent_id}/run", response_model=AutonomousAgentResponse)
async def run_autonomous_agent_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    schedule_autonomous_agent_run(agent.id)
    await db.refresh(agent)
    return agent


@router.post("/autonomous-agents/{agent_id}/pause", response_model=AutonomousAgentResponse)
async def pause_autonomous_agent_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    updated = await autonomous_agents_service.pause_autonomous_agent(db, agent.id)
    await db.commit()
    await db.refresh(updated)
    return updated


@router.post("/autonomous-agents/{agent_id}/resume", response_model=AutonomousAgentResponse)
async def resume_autonomous_agent_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    schedule_autonomous_agent_run(agent.id)
    await db.refresh(agent)
    return agent


@router.post("/autonomous-agents/{agent_id}/stop", response_model=AutonomousAgentResponse)
async def stop_autonomous_agent_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    updated = await autonomous_agents_service.stop_autonomous_agent(db, agent.id)
    await db.commit()
    await db.refresh(updated)
    return updated


@router.get("/autonomous-agents/{agent_id}/status", response_model=AgentStatusResponse)
async def get_autonomous_agent_status_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    return await autonomous_agents_service.get_agent_status(db, agent.id)


@router.get("/autonomous-agents/{agent_id}/cost", response_model=AgentCostResponse)
async def get_autonomous_agent_cost_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    return await autonomous_agents_service.get_agent_cost(db, agent.id)


# --------------------------------------------------------------- planning

@router.get("/autonomous-agents/{agent_id}/plans", response_model=list[AgentPlanResponse])
async def list_agent_plans_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    return await autonomous_agents_service.list_agent_plans(db, agent.id)


@router.get("/autonomous-agents/{agent_id}/plans/{plan_id}", response_model=AgentPlanResponse)
async def get_agent_plan_endpoint(
    plan_id: uuid.UUID, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db),
):
    agent, _caller = agent_ctx
    plan = await autonomous_agents_service.get_agent_plan(db, plan_id)
    if plan is None or plan.agent_id != agent.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    return plan


@router.get("/autonomous-agents/{agent_id}/plans/{plan_id}/steps", response_model=list[AgentStepResponse])
async def list_agent_steps_endpoint(
    plan_id: uuid.UUID, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db),
):
    agent, _caller = agent_ctx
    plan = await autonomous_agents_service.get_agent_plan(db, plan_id)
    if plan is None or plan.agent_id != agent.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    return await autonomous_agents_service.list_agent_steps(db, plan_id)


# --------------------------------------------------------------- memory

@router.get("/autonomous-agents/{agent_id}/memory", response_model=list[AgentMemoryResponse])
async def get_agent_memory_endpoint(
    memory_type: str | None = None, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db),
):
    agent, _caller = agent_ctx
    return await autonomous_agents_service.get_agent_memory(db, agent.id, memory_type)


@router.post("/autonomous-agents/{agent_id}/memory", response_model=AgentMemoryResponse, status_code=status.HTTP_201_CREATED)
async def add_agent_memory_endpoint(
    payload: AgentMemoryCreateRequest, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db),
):
    agent, _caller = agent_ctx
    memory = await autonomous_agents_service.add_agent_memory(db, agent.id, payload.content, payload.memory_type, payload.importance)
    await db.commit()
    await db.refresh(memory)
    return memory


@router.delete("/autonomous-agents/{agent_id}/memory/{memory_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_agent_memory_endpoint(
    memory_id: uuid.UUID, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db),
):
    _agent, _caller = agent_ctx
    await autonomous_agents_service.delete_agent_memory(db, memory_id)
    await db.commit()


# --------------------------------------------------------------- collaboration

@router.post("/autonomous-agents/{agent_id}/collaborate", response_model=AgentCollaborationResponse, status_code=status.HTTP_201_CREATED)
async def collaborate_endpoint(
    payload: AgentCollaborationRequest, agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db),
):
    agent, _caller = agent_ctx
    collaboration = await autonomous_agents_service.collaborate_agents(db, agent.id, payload.collaborator_agent_id, payload.task)
    await autonomous_agents_service.accept_collaboration(db, collaboration.id)
    await autonomous_agents_service.execute_collaboration(db, collaboration.id)
    await db.commit()
    await db.refresh(collaboration)
    return collaboration


@router.get("/autonomous-agents/{agent_id}/collaborations", response_model=list[AgentCollaborationResponse])
async def list_collaborations_endpoint(agent_ctx: tuple[AutonomousAgent, OrganizationMember] = Depends(require_autonomous_agent_member), db: AsyncSession = Depends(get_db)):
    agent, _caller = agent_ctx
    return await autonomous_agents_service.list_collaborations(db, agent.id)
